cmd/tools/genrpcserverinterceptors/main.go

```diff
@@ -0,0 +1,300 @@
+
+const maxRequestDepth = 5
+
+type (
+	requestData struct {
+		Type string
+
+		WorkflowIdGetter string
+		RunIdGetter      string
+		TaskTokenGetter  string
+	}
+
+	grpcServerData struct {
+		Server   string
+		Imports  []string
+		Requests []requestData
+	}
+)
+
+var (
+	// List of types for which Workflow tag getters are generated.
+	grpcServers = []reflect.Type{
+		reflect.TypeOf((*workflowservice.WorkflowServiceServer)(nil)).Elem(),
+		reflect.TypeOf((*adminservice.AdminServiceServer)(nil)).Elem(),
+		reflect.TypeOf((*historyservice.HistoryServiceServer)(nil)).Elem(),
+		reflect.TypeOf((*matchingservice.MatchingServiceServer)(nil)).Elem(),
+	}
+
+	// Only request fields that end with these suffixes are eligible for deeper inspection.
+	rootLevelFieldNameSuffixes = []string{
+		"Request",
+		"Completion",
+		"Ref",
+		"ParentExecution",
+		"WorkflowState", // for ReplicateWorkflowStateRequest
+	}
+
+	// These types have task_token field, but it is not of type *tokenspb.Task and doesn't have Workflow tags.
+	excludeTaskTokenTypes = []reflect.Type{
+		reflect.TypeOf((*workflowservice.RespondQueryTaskCompletedRequest)(nil)),
+		reflect.TypeOf((*workflowservice.RespondNexusTaskCompletedRequest)(nil)),
+		reflect.TypeOf((*workflowservice.RespondNexusTaskFailedRequest)(nil)),
+	}
+
+	executionGetterT = reflect.TypeOf((*interface {
+		GetExecution() *commonpb.WorkflowExecution
+	})(nil)).Elem()
+
+	workflowExecutionGetterT = reflect.TypeOf((*interface {
+		GetWorkflowExecution() *commonpb.WorkflowExecution
+	})(nil)).Elem()
+
+	taskTokenGetterT = reflect.TypeOf((*interface {
+		GetTaskToken() []byte
+	})(nil)).Elem()
+
+	workflowIdGetterT = reflect.TypeOf((*interface {
+		GetWorkflowId() string
+	})(nil)).Elem()
+
+	runIdGetterT = reflect.TypeOf((*interface {
+		GetRunId() string
+	})(nil)).Elem()
+)
+
+func main() {
+	outPathFlag := flag.String("out", ".", "path to write generated files")
+	licenseFlag := flag.String("copyright_file", "../../../LICENSE", "path to license to copy into header")
+	flag.Parse()
+
+	licenseText := readLicenseFile(*licenseFlag)
+
+	for _, grpcServerT := range grpcServers {
+		callWithFile(generateWorkflowTagGetters, grpcServerT, *outPathFlag, licenseText)
+	}
+}
+
+func generateWorkflowTagGetters(w io.Writer, grpcServerT reflect.Type) {
+
+	writeGrpcServerData(w, grpcServerT, `
+package logtags
+
+import (
+{{- range .Imports}}
+	{{printf "%q" .}}
+{{- end}}
+	"go.temporal.io/server/common/log/tag"
+)
+
+func (wt *WorkflowTags) extractFrom{{.Server}}Request(req any) []tag.Tag {
+	switch r := req.(type) {
+	{{- range .Requests}}
+	case {{.Type}}:
+	{{- if or .TaskTokenGetter .WorkflowIdGetter .RunIdGetter}}
+		{{- if .TaskTokenGetter}}
+		return wt.fromTaskToken(r.{{ .TaskTokenGetter}})
+		{{- else}}
+		return []tag.Tag{
+		{{if .WorkflowIdGetter}}	tag.WorkflowID(r.{{.WorkflowIdGetter}}),
+		{{end -}}
+		{{if .RunIdGetter}}	tag.WorkflowRunID(r.{{.RunIdGetter}}),
+		{{end -}}
+		}
+		{{- end}}
+	{{- else}}
+		return nil
```

Comment:
should we not generate a `case` for these at all? we have the `default: return nil` so they're not needed

Reply:
I considered it, and I prefer to explicitly list request types that don't have workflow tags.